### ISSUE TITLE
Blocked access to a diagnosis room is not resolved by making a path available

### ISSUE THREAD

**Issue by Alberth289346** [CorsixTH/CorsixTH#2090]
Basically a bug except it involves blocked room access and thus is part of #826 

If a diagnosis room is not reachable due to a blocked path, the game perceives that as not having such a room. (It performs a `findroomNearby` call, and obviously a path finder is not going to find such rooms.) 

You get a fax to build the room, and the patient involved becomes blocked on a 'built room' event.

Unblocking the path does nothing for the blocked patient (new patients will find the room properly).

**Comment by Go-rom** [comment]
Hello,

i'm encountering a similar issue except it doesn't seem to involve a blocked path.
In the same game, sometimes the patient find the way to to the diganosis room and sometimes not, i don't know why.

It seems to be related to some specific diseases like invisibility.
All invisible patients will never find the diagnosis room.

In order to provide more information, i turned on debug mode, hoping to get a relevant log file when the invisible patient doesn't find the diagnosis room but nothing. If there is any way to do that, please tell me.

**Comment by lewri** [comment]
From some tinkering, if you go into editing the affected room, then backing out, it sometimes seems to give the human a poke.
You could also try sending the person to the receptionist to see if it resolves.

Just noting potential workarounds while it's looked into.

**Comment by mugmuggy** [comment]
Yeah that the build callback from the seek_room action, kicking into gear. There is a scenario where you can (I might have fixed it previously), edit a room when you have say a single treatment room and an emergency patient. Edit the room but abort, and it would have them paused waiting for that callback (unless they vomit which interrupts and requeues the action). So this was not all events firing when they should.

The blocked off areas issue is more of the way that's working, and when comparing to TH. TH just doesn't do a full path check for accessibility when placing a rooms door. So one room can be placed in a blocked off area. Subsequent rooms being attempted to be placed in that blocked off area, fails, as the previous rooms door is not accessible to other doors. In a large enough area, you maybe also able to manipulate another room being placed in TH, by ensuring that the path way points exceeds the internal capacity of the path navigation.

Likewise I think if there are objects like bench/drinks machine, once the area is blocked you can't place anymore, because of this check, so it does allow blocked off areas constructed in particular ways.

There was also a previous issue around blocked paths not allowing staff to rest and that too would not alleviate this issue as they don't have callbacks either.

**Comment by Alberth289346** [comment]
Perhaps the simpler solution is to have humanoids idling, and attempt path finding themselves every now and then while waiting.

Since not finding a path is relatively expensive, as an optimization we could have a pessimistic path finder for this purpose that caches failed requests for some time, and returns "not available" to the same requests until the cached time runs out.

This would make at least the build-room callbacks obsolete.

**Comment by mugmuggy** [comment]
That was my thinking too, for seek_room reinsert an idle with a timer to restart the seek_room, as long as faxes remain suppressed after initial creation, that polling vs event notification is the only difference, but then again you could notify all patients not just those will callbacks registered and have the patient/humanoids work it out

**Comment by Alberth289346** [comment]
> ,,, but then again you could notify all patients not just those will callbacks registered and have the patient/humanoids work it out

Imho that's another good option, the only problem would be when to send out such a notification. Clearly the current activation points of the callbacks isn't enough, not sure you'd want it every time you build or remove some room or some item.

On the other hand, could be nice, put down an item that blocks a path, and all patients needing that path know instantly :)
I think it could easily get out of hand wrt required cpu time.

**Comment by ARGAMX** [comment]
Currently, in CorsixTH, regular players can’t build a blocked room.
Technically, you can only do this by disabling the checks for the blocked area via the developer menu. 
But I don’t consider the states that results from disabling this checks are valid. You turn off these checks at your own risk.

The original game also didn’t want players to build blocked rooms. 
But there was a bug that still allowed the user to build 1 such room. 
That’s exactly what mugmuggy is talking about.
But it was impossible to build a second one blocked room. 
And after building the first such room, the game started to glitch, until user remove the block.
From this, I conclude that Bullfrog didn’t intend to allow rooms to be blocked at all, and it’s only possible to block one room due to a bug. 

Based on all this, the issue described essentially not valid. So there’s nothing to fix.
